blossom: Initial Blossom support

danieldaquino opened the pull request:
### Description

This commit adds initial Blossom support via:
1. the implementation of Blossom-related event kinds and tags.
2. the implementation of a new `blossom` crate containing models, as well as a Blossom client implementation.

Closes https://github.com/rust-nostr/nostr/issues/800

### Notes to the reviewers

- The `blossom_auth` event builder function was defined inside the blossom crate with a special extension instead of the `nostr` crate because it depends on Blossom specific data structures that probably do not belong in the `nostr` crate.
- The event kind and special `server` tag were defined inside the `nostr` crate because those enums cannot be externally extended.
- I tried to follow the import order conventions outlined in the coding style guidelines, but the `just check` or `just precommit` seems to have overridden that order. Since these commands are enforced by CI, I gave preference to the order stipulated by the scripts.
- Testing performed focused mostly on manual testing instead of automated unit testing because the models themselves are very simple, and the client code requires interacting with a real Blossom server.
- This is my first contribution to this repo. I tried to follow the guidelines as much as I could, but let me know if I missed anything.

### Testing notes

**Version used for this repo:** d36a58884eadeee17de835e1a5342b9bbf5679bf
**Blossom server used:** Tested against https://github.com/0xtrr/cherry-server on version ab9bc434ae863b7f2773d0f76956beca9e4fd657
**Setup:** 
- cherry-server running with this configuration:
```toml
# Basic settings
database_directory = './db'
files_directory = './blobs'
server_url = 'http://laptop:3030' # Set to entire URL without trailing /. Used to set the URL in a blob descriptor.
host = '0.0.0.0:3030' # hostname:port to bind on

# Get blob settings
[get]
require_auth = true # Require authorization to fetch blob

# Upload blob settings
[upload]
enabled = true # Enable/disable uploads
max_size = 50 # In MB

# Upload blob pubkey filter
[upload.public_key_filter]
enabled = false 
mode = "Blacklist" # Specify if pubkey list is a whitelist or blacklist
public_keys = [
]

# Upload blob mimetype filter
[upload.mimetype_filter]
enabled = true
mode = "Whitelist" # Specify if mimetype list is a whitelist or blacklist
mime_types = [
    "image/png",
    "image/jpg",
    "image/jpeg",
    "image/webp",
    "image/gif",
    "video/mp4",
    "text/plain",
    "application/octet-stream"
]

# List blobs settings
[list]
require_auth = true # Require authorization for listing blobs

# Mirror blobs settings
[mirror]
enable = false # Enable /mirror endpoint
```
- server accessible from a non-localhost address (to make the test more realistic)

**Steps:**
1. Run the "integration test" code example:
```sh
KEY=$(nak key generate)
cargo run --example integration_test -- --server http://laptop:3030 --private-key $KEY
```
2. Read the output, make sure all operations seem to be working
3. Run other examples manually, using a real file, and verify functionality. Reference shell code of commands that were run:
```sh
KEY=$(nak key generate)
PUBKEY=$(nak key public $KEY)
cargo run --example upload -- --server http://laptop:3030 --file ~/Downloads/picture.png --content-type "image/png" --private-key $KEY
cargo run --example list -- --server http://laptop:3030 --private-key $KEY --pubkey $PUBKEY		# Verified uploaded image is listed
cargo run --example download -- --server http://laptop:3030 --sha256 69fbdd273b3dc430853381f8a782f53dd3ceceef050d8e760b1944a21a5a8f18 --private-key $KEY
mv 69fbdd273b3dc430853381f8a782f53dd3ceceef050d8e760b1944a21a5a8f18 69fbdd273b3dc430853381f8a782f53dd3ceceef050d8e760b1944a21a5a8f18.png
open 69fbdd273b3dc430853381f8a782f53dd3ceceef050d8e760b1944a21a5a8f18.png	# Visually verified the image is the same
cargo run --example delete -- --server http://laptop:3030 --sha256 69fbdd273b3dc430853381f8a782f53dd3ceceef050d8e760b1944a21a5a8f18 --private-key $KEY
cargo run --example list -- --server http://laptop:3030 --private-key $KEY --pubkey $PUBKEY		# Verified image is no longer listed
```

### Checklist

* [x] I followed the [contribution guidelines](https://github.com/rust-nostr/nostr/blob/master/CONTRIBUTING.md)
* [x] I ran `just precommit` or `just check` before committing _(Note 1)_
* [ ] I updated the [CHANGELOG](https://github.com/rust-nostr/nostr/blob/master/CHANGELOG.md) (if applicable) _(Note 2)_

_(Note 1)_ The scripts seem to have formatted my code, but I ran into an issue when running these scripts:
```
warning: secp256k1-sys@0.10.1: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
warning: secp256k1-sys@0.10.1: 1 error generated.
```

Even though I already have the wasm target:
```
% rustup target add wasm32-unknown-unknown
info: component 'rust-std' for target 'wasm32-unknown-unknown' is up to date
```

Since I have not made any substantial changes to other crates, I am assuming that there is a low risk the changes have broken existing crates, and this will probably be automatically checked by CI again before merging anyways?

_(Note 2)_ I believe it is not applicable, because otherwise I would have to increase the version number and add only one changelog item — which does not seem to match previous changelog version entries. Please let me know if this needs to be done.

yukibtc commented on crates/blossom/Cargo.toml:
```diff
@@ -0,0 +1,24 @@
+[package]
```
Rewrite this to:

```toml
[package]
name = "nostr-blossom"
version = "0.41.0"
edition = "2021"
description = "A library for interacting with the Blossom protocol"
authors.workspace = true
homepage.workspace = true
repository.workspace = true
license.workspace = true
readme = "README.md"
rust-version.workspace = true
keywords = ["nostr", "blossom"]

[dependencies]
base64.workspace = true
nostr = { workspace = true, features = ["std"] }
reqwest = { workspace = true, default-features = false, features = ["json", "rustls-tls", "socks"] }
serde = { workspace = true, features = ["derive"] }
serde_json.workspace = true

[dev-dependencies]
clap = { version = "4.0", features = ["derive"] }
tokio = { workspace = true, features = ["full"] }
```

The `blossom` crate is already used for another project, so I would rename it to `nostr-blossom`. The `bitcoin_hashes` is already re-exported from the `nostr` crate so can be removed (it's re-exported as `nostr::hashes`). The `base64` and `reqwest` deps aren't currently in the workspace, but since are used also in the `nostr` crate we can add them to the workspace.

And, can the `blossom` feature be removed? I saw that is never used

yukibtc commented on crates/blossom/src/lib.rs:
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) 2023-2025 Rust Nostr Developers
+// Distributed under the MIT software license
+
+#![forbid(unsafe_code)]
```
Sorry for nitpicking... just to keep the same format with the other crates, move the module-level doc above the crate-level attributes:

```rust
//! This crate implements the Blossom protocol for decentralized content storage and retrieval.
//!
//! The Blossom protocol defines a standard for storing and retrieving blobs (binary large objects)
//! in a decentralized manner, using the Nostr protocol for authorization and discovery.
//!
//! <https://github.com/hzrd149/blossom>

#![forbid(unsafe_code)]
#![warn(missing_docs)]
#![warn(rustdoc::bare_urls)]
#![warn(clippy::large_futures)]
```

yukibtc commented on crates/nostr/src/event/tag/kind.rs:
```diff
@@ -125,6 +125,8 @@ pub enum TagKind<'a> {
     /// <https://github.com/nostr-protocol/nips/blob/master/C0.md>
     Runtime,
+    /// Server
+    Server,
```
Miss the conversion from a string to TagKind in `impl<'a> From<&'a str> for TagKind<'a>`

yukibtc commented on crates/nostr/src/event/tag/standard.rs:
```diff
@@ -188,6 +188,7 @@ pub enum TagStandard {
         key: String,
         iv: String,
     },
+    Server(String),
```
Add also this line in the `internal_parse`:

```rust
if tag_len == 2 {
    let tag_1: & str = tag[1].as_ref();

    return match tag_kind {
        // ...
        TagKind::Server => Ok( Self::Server(tag_1.to_string())), // Add this
        TagKind::Summary => Ok(Self::Summary(tag_1.to_string())),
        // ...
    }
}
```

yukibtc reviewed (changes requested):
Thank you! I've flagged some changes to do. Sorry for nitpicking...

yukibtc commented:
> warning: secp256k1-sys@0.10.1: error: unable to create target: 'No available targets are compatible with triple "wasm32-unknown-unknown"'
warning: secp256k1-sys@0.10.1: 1 error generated.

Are you on macOS? This seems to happen on macOS (https://github.com/rust-nostr/nostr/pull/836#issuecomment-2812098604)

danieldaquino commented:
> Thank you! I've flagged some changes to do. Sorry for nitpicking...

Thank you for your quick and thorough review! Rust is not the main language I usually work with, so it was nice to learn how I can write better and cleaner Rust code.

I have addressed your comments and updated the PR. I ran `just check` and everything that could run on my setup seems to have passed. (Except for the wasm portion and everything after that, due to that macOS-related issue). I also re-ran the `integration_test` example program to ensure it still works as expected.

> Are you on macOS? This seems to happen on macOS (https://github.com/rust-nostr/nostr/pull/836#issuecomment-2812098604)

Yes! I tried to follow the steps in that comment, and I was able to identify some issues in the new crate on the Wasm target, which I have fixed in this new version of the PR. After fixing the Wasm build for blossom-nostr, I ran into issues building the crate `mio-1.0.2`, which seems unrelated to the new crate, so I am assuming it's another dev environment issue.

---

In this new version, I also added the changelog entry (I found out what I needed to do by looking at another PR).


Please let me know if you have any other concerns or suggestions.

Thank you!

yukibtc commented:
Thanks! I've rebased, fixed the conflicts and merged at eff00fcdd74f4eb50048176b12362785390a6865
